Feature/livedisplay improvements

Quallenauge opened the pull request:
Make livedisplay driver more robust.
Add changes to apply DSI commands (e.g. cabc) to both DSI panel instances.

Commit:
video: livedisplay: Allow setting DSI link state via DT

 * We need to send commands in LP mode for some hardware.

Change-Id: I96dacea6bcc52960f9f759981984752180a33464

kerneltoast committed:
mdss: livedisplay: Fix memory leaks in mdss_livedisplay_update()

parse_dsi_cmds() allocates two blocks of memory and stores their addresses
in a stack-allocated variable, but they are never freed, so when the
function exits, all references to them are gone.

Free the allocated memory after using it in order to fix the memory leaks.

Change-Id: Ie574848e2429167fc38ed39975feb3df68ed2aed
Signed-off-by: Sultanxda <sultanxda@gmail.com>

javelinanddart committed:
mdss: livedisplay: Don't allow non-contiguous presets

* We currently allow having preset numbers that aren't
  continuous, but the rest of the preset code isn't
  designed for this, so disallow this
* This also speeds up parse_dt by breaking out of the
  loop early if we don't have any presets instead of
  looking for all 10 presets even if the first isn't there

Change-Id: I3a48b8ef027afcdbd27ba92bc43f5438c0f249f2

Quallenauge committed:
mdss: livedisplay: Correct mdss_dsi_panel_cmds_send function signature.

Function mdss_dsi_panel_cmds_send takes a u32 field for setting
flags. This is needed to specific tx/rx direction for example.

Quallenauge committed:
dts: mdss: Move CABC initialization to dedicated dts parameters.

Once the nodes
cm,mdss-livedisplay-cabc-cmd
cm,mdss-livedisplay-cabc-ui-value
cm,mdss-livedisplay-cabc-image-value
cm,mdss-livedisplay-cabc-video-value
exists, they where used by the livedisplay lineage os driver.

Quallenauge committed:
dts: mdss: dsi: panel on command support after te enable

Few DSI panels has requirement of receiving display on command
after receiving sleep out command and pixel stream data. This
can be achieved by sending panel post on command followed by
timing engine enable. This change adds the support for such
kind of DSI panels.

This setting is found on all newer caf trees for this display.

(Source commit: msm: mdss: add panel on command support after te enable )

Quallenauge committed:
video: mdss: livedisplay: Add support for secondary panels.

When using dual displays, only the first controller
DSI Ctrl name = MDSS DSI CTRL->0
Panel Name = nt35597 video mode dual dsi panel without DSC
one is connected to a framebuffer. At the initialization of the framebuffer
the sysfs nodes of the livedisplay are created. If this is successful,
the framebuffer ( struct msm_fb_data_type *mfd ) is assigned to the livedisplay
structures.

For the second controller
DSI Ctrl name = MDSS DSI CTRL->1
Panel Name = nt35597 video mode dual dsi panel without DSC
this isn't the case. There's no framebuffer assignment due missing
sysfs creation call.

This is caused at the registration of the framebuffer.
At mdss_register_panel() there is a call to add an
extra panel to the same fb via call of function
mdss_fb_register_extra_panel().

It sets the panel to the address of itself to
struct mdss_panel_data fb_pdata->next = pdata;
For the extra panel, there's no separate sysfs creation performed.

At the livedisplay update routine, it is checked if there's an
value at struct mdss_panel_data fb_pdata->next assigned.
If so, the dedicated controller is determined and the dsi commands
are send to that controller in addition to the original controller.

Gabriel85 commented:
Good! What about the image "enhancement" features?

Gabriel85 commented on arch/arm/boot/dts/qcom/YTX703.dtsi:
```diff
@@ -282,11 +282,19 @@
 			15 01 00 00 01 00 02 f7 01
 			39 01 00 00 01 00 05 ff aa 55 a5 00
 			15 01 00 00 01 00 02 62 01
-			15 01 00 00 01 00 02 55 00 //CABC mode (0:off, 1:UI, 2:picture, 3:video), p.229
```
Don't remove this please

Quallenauge replied:
Is there a reason for?
I see a drawback when letting this within the command sequence (But I'm not sure atm.).
This needs to be tested.

Gabriel85 replied:
Yes, to keep CABC disabled, in chase we ever remove LiveDisplay. As far as I remember it is enabled by default. But depends on if your commits, reapply the settings every time the Display is turned on, or if they only are applied once.

Quallenauge replied:
The idea is good, but fails when the second display panel instances comes up.
Display on sequence:
1. The first initializes the CABC via the livedisplay driver.
2. Then the second comes up and overwrites the previously set value with the "disabled" state.
Result: When the display is switched on, then the CABC will be not active in any case.

Quallenauge committed:
video: livedisplay: Use static lock mutex.

Updating of the panels seems to be performed asynchronly.
The livedisplay driver tries to perform it's actions by
locking a mutext of the panel instance specific
struct mutex lock.

Since both panels are using the same hw, the lock is useless
for actions which are performed in parallel on both panel instances,
since they are using their own lock field for locking.

To workaround this, the lock is made global, which is the drawback
that all panels, which are handled by the livedisplay driver,
are using this single lock which can be causes some performance
issues which can be ignored, since the events
(display on / setRGB / ...) appear with a low rate.

Gabriel85 commented:
Mhh, but otherwise will never be disabled, unless you set it manually in Livedisplay? Right now CABC is enabled until you switch your display off and on again, after boot. This is why I thought, that is isn't fully disabled. Not good..

On 23 July 2018 09:24:05 CEST, Quallenauge <notifications@github.com> wrote:
>Quallenauge commented on this pull request.
>
>
>
>> @@ -282,11 +282,19 @@
> 			15 01 00 00 01 00 02 f7 01
> 			39 01 00 00 01 00 05 ff aa 55 a5 00
> 			15 01 00 00 01 00 02 62 01
>-			15 01 00 00 01 00 02 55 00 //CABC mode (0:off, 1:UI, 2:picture,
>3:video), p.229
>
>The idea is good, but fails when the second display panel instances
>comes up.
>Display on sequence:
>1. The first initializes the CABC via the livedisplay driver.
>2. Then the second comes up and overwrites the previously set value
>with the "disabled" state.
>Result: When the display is switched on, then the CABC will be not
>active in any case.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#discussion_r204302264

Gabriel85 commented:
I looked into this "Image Enhancement" stuff, and turns out the documentation I have is mostly unusable. I tried out those things mentioned there, but they didn't seem to work or the difference is so small that I can't see it. The only other Livedisplay implementation of a similar controller, implements also only CABC. Therefore I assume it doesn't support any. Although it supports colour inversing..

But there is a bigger problem: those CABC is totally wrong set up in the first place! Not only it doesn't work, it does the complete opposite! Instead of dimming the display on bright scenes, it makes it brighter (please confirm). Total fail...
Seems that they didn't set up the PWM duty cycle according to the backlight controller needs. But I don't have the documentation to fix it. Only workaround I can imagine at the moment is to switch from the current I2C+PWM backlight control, to a pure PWM solution over DSI. Therefore only the panel itself will be able to directly manipulate the brightness.

Does anybody know if the stockrom offers any hardware image stuff?
And those it supports several CABC modes, or just ON/OFF?

On 22 July 2018 14:16:46 CEST, Quallenauge <notifications@github.com> wrote:
>Make livedisplay driver more robust.
>Add changes to apply DSI commands (e.g. cabc) to both DSI panel
>instances.
>You can view, comment on, or merge this pull request online at:
>
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17
>
>-- Commit Summary --
>
>  * dts: mdss: dsi: panel on command support after te enable
>  * mdss: livedisplay: Fix memory leaks in mdss_livedisplay_update()
>* mdss: livedisplay: Correct mdss_dsi_panel_cmds_send function
>signature.
>  * mdss: livedisplay: Don't allow non-contiguous presets
>  * video: livedisplay: Allow setting DSI link state via DT
>  * mdss: livedisplay: Move update before post panel on dsi cmds.
>  * dts: mdss: Move CABC initialization to dedicated dts parameters.
>  * video: mdss: livedisplay: Add support for secondary panels.
>
>-- File Changes --
>
>    M arch/arm/boot/dts/qcom/YTX703.dtsi (14)
>    M drivers/video/msm/mdss/mdss_dsi_panel.c (4)
>    M drivers/video/msm/mdss/mdss_livedisplay.c (31)
>    M drivers/video/msm/mdss/mdss_livedisplay.h (2)
>
>-- Patch Links --
>
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17.patch
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17.diff
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17

Quallenauge commented:
Regardless the issues you mentioned, I have to review my commits.
As it seems I have a lessen learned, always to make a ccache clean in case issues occurs.
Please don't use them right now.

Gabriel85 commented:
Just to make it clear, the CABC issues at NOT caused by you, they are Lenovo's fault. And probably affect the stock ROM, too.
But you might want to wait working on that stuff, until I found a fix for it. If we can't fix CABC, there is absolutely no point in using it, not only it is nasty at the moment, it also increased battery usage.. 

On 23 July 2018 22:16:28 CEST, Quallenauge <notifications@github.com> wrote:
>Regardless the issues you mentioned, I have to review my commits.
>As it seems I have a lessen learned, always to make a ccache clean in
>case issues occurs.
>Please don't use them right now.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407186645

Gabriel85 commented:
OK, turned out the Lenovo guys bought the wrong backlight controller! The LP8557 doesn't support the current implementation. Although to be fair the  documentation makes it seem it those. There for my proposed workaround it the only working solution, and should have been implemented in the first place. Amazing how much incompetence there was during the development of our tablet...
So we have to change the implementation of the backlight control in the kernel and device repo. I gonna open a separate pull-request or issue there.

And concerning that Livedisplay stuff: it seems awfully complicated just to set a certain DSI command (which basically also could be done by a simple script). Any reasonable implementation must be doing two things:
1. ASAP during boot disable CABC (probably can be done while the display is on, but need to confirm that)
2. if the sys file entry is changed, immediately send the corresponding DSI command + redo it every time after the display turned on

But if we implement everything correctly, CABC will actually be a useful feature and will work way better than in the stock ROM. At least my first test are very promising, in video mode (strongest) it is already way better than the stock implementation in UI mode (weakest). And it actually decreases power consumption, instead of increasing it...

On 23 July 2018 17:41:41 CEST, Gabriel Speranza <Gabriel@Speranza.me> wrote:
>I looked into this "Image Enhancement" stuff, and turns out the
>documentation I have is mostly unusable. I tried out those things
>mentioned there, but they didn't seem to work or the difference is so
>small that I can't see it. The only other Livedisplay implementation of
>a similar controller, implements also only CABC. Therefore I assume it
>doesn't support any. Although it supports colour inversing..
>
>But there is a bigger problem: those CABC is totally wrong set up in
>the first place! Not only it doesn't work, it does the complete
>opposite! Instead of dimming the display on bright scenes, it makes it
>brighter (please confirm). Total fail...
>Seems that they didn't set up the PWM duty cycle according to the
>backlight controller needs. But I don't have the documentation to fix
>it. Only workaround I can imagine at the moment is to switch from the
>current I2C+PWM backlight control, to a pure PWM solution over DSI.
>Therefore only the panel itself will be able to directly manipulate the
>brightness.
>
>Does anybody know if the stockrom offers any hardware image stuff?
>And those it supports several CABC modes, or just ON/OFF?
>
>On 22 July 2018 14:16:46 CEST, Quallenauge <notifications@github.com>
>wrote:
>>Make livedisplay driver more robust.
>>Add changes to apply DSI commands (e.g. cabc) to both DSI panel
>>instances.
>>You can view, comment on, or merge this pull request online at:
>>
>>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17
>>
>>-- Commit Summary --
>>
>>  * dts: mdss: dsi: panel on command support after te enable
>>  * mdss: livedisplay: Fix memory leaks in mdss_livedisplay_update()
>>* mdss: livedisplay: Correct mdss_dsi_panel_cmds_send function
>>signature.
>>  * mdss: livedisplay: Don't allow non-contiguous presets
>>  * video: livedisplay: Allow setting DSI link state via DT
>>  * mdss: livedisplay: Move update before post panel on dsi cmds.
>>  * dts: mdss: Move CABC initialization to dedicated dts parameters.
>>  * video: mdss: livedisplay: Add support for secondary panels.
>>
>>-- File Changes --
>>
>>    M arch/arm/boot/dts/qcom/YTX703.dtsi (14)
>>    M drivers/video/msm/mdss/mdss_dsi_panel.c (4)
>>    M drivers/video/msm/mdss/mdss_livedisplay.c (31)
>>    M drivers/video/msm/mdss/mdss_livedisplay.h (2)
>>
>>-- Patch Links --
>>
>>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17.patch
>>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17.diff
>>
>>-- 
>>You are receiving this because you are subscribed to this thread.
>>Reply to this email directly or view it on GitHub:
>>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17

Quallenauge commented:
What did you try exactly?

> And concerning that Livedisplay stuff: it seems awfully complicated just to set a certain DSI command (which basically also could be done by a simple script).

I don't see any advantage over an self baked script.

>There for my proposed workaround

Which one? Do I miss something?

>1. ASAP during boot disable CABC (probably can be done while the display is on, but need to confirm that)

That's possible. At least I saw effects while setting the sysfs node.

> 2. if the sys file entry is changed, immediately send the corresponding DSI command + redo it every time after the display turned on

The first part is exactly what the implementation does. 

> redo it every time after the display turned on

I don't understand.

Gabriel85 commented:
Didn't say that it isn't already doing that, but simply writing the sysfs entry with the inits during boot probably would be a nice solution.
I mean that the display controls the backlight, purely over PWM, instead of the current (non-working) PWM+I2C way.
I give more information about it, after I have confirmed that it is actually working as intended.

On 24 July 2018 08:43:35 CEST, Quallenauge <notifications@github.com> wrote:
>What did you try exactly?
>
>> And concerning that Livedisplay stuff: it seems awfully complicated
>just to set a certain DSI command (which basically also could be done
>by a simple script).
>
>I don't see any advantage over an self baked script.
>
>>There for my proposed workaround
>
>Which one? Do I miss something?
>
>>1. ASAP during boot disable CABC (probably can be done while the
>display is on, but need to confirm that)
>
>That's possible. At least I saw effects while setting the sysfs node.
>
>> 2. if the sys file entry is changed, immediately send the
>corresponding DSI command + redo it every time after the display turned
>on
>
>The first part is exactly what the implementation does. 
>
>> redo it every time after the display turned on
>I don't understand.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407298893

Quallenauge commented:
With I2C you mean the brightness registers?

Gabriel85 commented:
Yup, basically we write the desired brightness into there at the moment. But the display requests another value at the same time via PWM. Than the backlight controller sets the actual brightness as a multiplication of both. Which doesn't really matter as long as one of those never changes. But with CABC enabled the display adjusts the brightness constantly, and as a result we get garbage..

On 24 July 2018 08:54:35 CEST, Quallenauge <notifications@github.com> wrote:
>With I2C you mean the brightness registers?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407301236

Quallenauge commented:
I'm curious if it is possible that when the backlight controller runs in PWM `<0x00>` mode and then let the display control the brightness completely via `/sys/class/leds/lcd-backlight/brightness` sysfs node or brightness dsi command ( (5300h) WRCTRLD: Write CTRL Display  ).

Gabriel85 commented:
Yeah something like that, but not with this sysfs node. I need to reconfigure the backlight and display controller and find out if there are any disadvantages, but luckily I have at least the full documentation for backlight controller. Although controlling it via PWM is surly inferior, but I hope the LP8557 is able to handle it as well as I2C.
Even more I hope it works with the standard PWM settings of the panel, since there is no way to modify those without proper documentation.. :(

On 24 July 2018 09:38:38 CEST, Quallenauge <notifications@github.com> wrote:
>I'm curious if it is possible that when the backlight controller runs
>in PWM `<0x00>` mode and then let the display control the brightness
>completely via `/sys/class/leds/lcd-backlight/brightness` sysfs node or
>brightness dsi command ( (5300h) WRCTRLD: Write CTRL Display  ).
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407311545

Quallenauge commented:
>  I need to reconfigure the backlight and display controller and find out if there are any disadvantages, but luckily I have at least the full documentation for backlight controller.

You mean this one?
http://www.ti.com/lit/ds/symlink/lp8557.pdf

Gabriel85 commented:
Exactly, although it falsely sounds like <0x11> mode would work with I2C and PWM input for CABC, but it seems they copied that part from another IC. But it can't work, since the panel has no way to know at which level the backlight is at the moment, therefore can't know which value it is the right one to slightly decrease brightness.
But I'm not sure yet if this is really the only problem, since it seems that CABC always increased the brightness, irrespectively of the current I2C setting.

On 24 July 2018 10:06:44 CEST, Quallenauge <notifications@github.com> wrote:
>>  I need to reconfigure the backlight and display controller and find
>out if there are any disadvantages, but luckily I have at least the
>full documentation for backlight controller.
>
>You mean this one?
>http://www.ti.com/lit/ds/symlink/lp8557.pdf
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407319116

Gabriel85 commented:
Bad news: turned out that it is probably a hardware bug. Even after "fixing" it, it still gets brighter when it should get darker. Removing the I2C setting even made it more pronounced. I even reset the controller completely, and only used settings, I know and still the same issue. Seems Lenovo is at least not to blame for that...
A similar panel from a Dell laptop had the same issue, and the only fix by them was to disable CABL completely. A pity since it would have been finally useful with the right settings. Theoretically the PWM polarity could be inversed, but I don't know how to set it up correctly. Also it would only be a partial fix, and I assume that not all devices are affected of the bug. So I guess it is probably the best to abandon all efforts to implement CABC for good. :(

On 24 July 2018 10:14:38 CEST, Gabriel Speranza <Gabriel@Speranza.me> wrote:
>Exactly, although it falsely sounds like <0x11> mode would work with
>I2C and PWM input for CABC, but it seems they copied that part from
>another IC. But it can't work, since the panel has no way to know at
>which level the backlight is at the moment, therefore can't know which
>value it is the right one to slightly decrease brightness.
>But I'm not sure yet if this is really the only problem, since it seems
>that CABC always increased the brightness, irrespectively of the
>current I2C setting.
>
>On 24 July 2018 10:06:44 CEST, Quallenauge <notifications@github.com>
>wrote:
>>>  I need to reconfigure the backlight and display controller and find
>>out if there are any disadvantages, but luckily I have at least the
>>full documentation for backlight controller.
>>
>>You mean this one?
>>http://www.ti.com/lit/ds/symlink/lp8557.pdf
>>
>>-- 
>>You are receiving this because you commented.
>>Reply to this email directly or view it on GitHub:
>>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407319116

Quallenauge commented:
I tried some more things and write down my findings:
### To recapitulate:
The panel is connected to the backlight controller. From the HW side of view, it is able to adjust the brightness of the backlight via pwm.

### DTS configuration
The modules are prepared for this scenario, where the panel controls the backlight. Therefore the panel
is informed about brightness values with the help of dsi commands. The relevant setting for this is
`qcom,mdss-dsi-bl-pmic-control-type = "bl_ctrl_dcs"; // control backlight via dsi commands`
Now the brightness levels have to be adjusted to a valid range.
```
qcom,mdss-dsi-bl-min-level = <0x01>;
qcom,mdss-dsi-bl-max-level = <0xFF>;
```
At the initial dsi command sequence, I moved the backlight relevant parts to the end to switch the backlight on, as late as possible. Also I set the initial brightness to zero, to hide the effect, that the backlight goes on, black screen and after a ~0.7 seconds period the picture is seen.

### Userspace
Now the brightness have to be controlled via the sysfs node:
`/sys/class/leds/lcd-backlight/brightness`

### Results
The display now knows about the target brightness and is able to perform the CABC calculation.
From my tests (and it is possible that they are very subjective) I found in cabc video mode (0x03),
that on bright screen content the backlight is normal, but on dark screen content it lowers the brightness a little bit.
 
### Diff for testing
```diff
diff --git a/arch/arm/boot/dts/qcom/YTX703.dtsi b/arch/arm/boot/dts/qcom/YTX703.dtsi
index 2a2f41aebd45..08c6a1725708 100644
--- a/arch/arm/boot/dts/qcom/YTX703.dtsi
+++ b/arch/arm/boot/dts/qcom/YTX703.dtsi
@@ -170,8 +170,9 @@
                compatible = "ti,lp8557";
                reg = <0x2c>;
                bl-name = "lcd-bl";
-               dev-ctrl = /bits/ 8 <0x07>; /* see lp8557 documentation */
-               init-brt = /bits/ 8 <0x40>; /* 25% brightness */
+               dev-ctrl = /bits/ 8 <0x00>; /* see lp8557 documentation, PWM mode only */
+               init-brt = /bits/ 8 <0x00>; /* 0% initial brightness */
+               pwm-period = <1000000>; /* value from documentation, doesn't seem to have an effect */
                 /* enable spread-spectrum boost clocking (reduces flickering): */
                rom_00h {
                        rom-addr = /bits/ 8 <0x00>;
@@ -273,6 +274,10 @@
                qcom,mdss-dsi-v-back-porch  = <24>;
                qcom,mdss-dsi-v-front-porch = <16>;
                qcom,mdss-dsi-v-pulse-width = <8>;
+               qcom,mdss-dsi-bl-pmic-control-type = "bl_ctrl_dcs"; // control backlight via dsi commands
+               /* Backlight controller operates in range 0<x<256 */
+               qcom,mdss-dsi-bl-min-level = <0x01>;
+               qcom,mdss-dsi-bl-max-level = <0xFF>;
                qcom,mdss-dsi-on-command = [
                        39 01 00 00 01 00 06 f0 55 aa 52 08 00
                        15 01 00 00 01 00 02 c0 0d
@@ -282,10 +287,11 @@
                        15 01 00 00 01 00 02 f7 01
                        39 01 00 00 01 00 05 ff aa 55 a5 00
                        15 01 00 00 01 00 02 62 01
-                       15 01 00 00 01 00 02 53 24
-                       15 01 00 00 01 00 02 51 80
                        05 01 00 00 78 00 02 11 00
-                       05 01 00 00 15 00 02 29 00];
+                       05 01 00 00 15 00 02 29 00
+                       15 01 00 00 01 00 02 51 00 // initial brightness set to 0
+                       15 01 00 00 01 00 02 53 24 // turn the lights on
+               ];
                qcom,mdss-dsi-panel-timings = [D5 32 22 00 60 64 26 34 29 03 04 00]; //MIPICLL 399
 
                /* CABC init commands / values */
```

Gabriel85 commented:
Great, so can we close that now? ;)

Quallenauge commented:
Ok, I'll open another one with reduced set of improvements and one which includes the conversation to 
/sys/class/leds/lcd-backlight/brightness.

Gabriel85 commented:
Which other improvements? But that conversation is surly not one, since direct I2C setting of the brightness settings is technically and practically superior to the usage of PWM. I actually plan to disable it completely, like stock does with PWM disabled.
I've also spend a lot of time on that CABL issue, and hate that it was in vain. But for me that LiveDisplay stuff is a dead end and I like to propose a separate implementation of that RGB PCC stuff (probably the same which is already included in Livedisplay, just without the now useless stuff), since it is now proven that we never ever will be able to use any other feature of it. We tried the best to get it to work, and I feel really sorry for it, but sometimes you have to admit defeat..

On 26 July 2018 07:18:21 CEST, Quallenauge <notifications@github.com> wrote:
>Ok, I'll open another one with reduced set of improvements and one
>which includes the conversation to 
>/sys/class/leds/lcd-backlight/brightness.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407979174

Quallenauge commented:
> since direct I2C setting of the brightness settings is technically and practically superior to the usage of PWM

I can't follow your argumentation.
Why is PWM not as good as I2C? I have to admit, I'm not an expert in electronics, so can you give me some insights please?
>  I actually plan to disable it completely, like stock does with PWM disabled.

I plan to make it a standard.

>And I like to propose a separate implementation of that RGB PCC

You can do it, but I think it's like reinvention the wheel. Maybe I'm wrong.

Gabriel85 commented:
1. Well as you probably already found out PWM uses a series of pulses and the average voltage determines the actual "value", therefore it is analogue, and it is not possible to deliver an exact value. The backlight controller converts this than back to a approximate digital value (and later in another PWM signal, but that doesn't matter). So it is better to send the intended digital value directly and the only way to actually set the backlight 100% accurately.
3.  The first Livedisplay commit was just a buggy mega-commit of several others. I prefer to use the original (much simpler and smaller) one instead. The fact that this thing had to be fixed that often and even contain(ed) memory leaks, speaks for itself, I think. The only stuff we ever can use makes about 2% of the complete Livedisplay code.

On 26 July 2018 07:40:49 CEST, Quallenauge <notifications@github.com> wrote:
>> since direct I2C setting of the brightness settings is technically
>and practically superior to the usage of PWM
>
>I can't follow your argumentation.
>Why is PWM not as good as I2C? I have to admit, I'm not an expert in
>electronics, so can you give me some insights please?
>>  I actually plan to disable it completely, like stock does with PWM
>disabled.
>
>I plan to make it a standard.
>
>>And I like to propose a separate implementation of that RGB PCC
>
>You can do it, but I think it's like reinvention the wheel. Maybe I'm
>wrong.
>
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407982578

Quallenauge commented:
1) Sure, there are issues. But to be honest, backlights in our devices have an operating range. Setting the brightness to 128 results in a different achieved effective brightness with your and my device. I think this can be tolerated, if the PWM sets the value to 128".1"/ 129".3333" or 127.9. The user won't notice that.

3) You can do that of course. And this is an never-ending discussion right now. So I end here with this: In my opinion, the existing driver can be improved instead of trim down or re-implement.

Gabriel85 commented:
1. I agree, but question here is not if it is not noticeable worse. The question is if it is any better, or at the very least not worse in any kind...
2. It's neither of those. It not that they guy who wrote the Livedisplay stuff has invented that code by himself. He just took it from somewhere else. So why don't use the original, if we don't need the other stuff?

And again: what improvements? It's not that I can't change my mind about that. In fact I already did several times, when I had a reason to! I was against it, than you convinced me, and now the facts just have changed. But unlike when the hole debate started we KNOW now that Livedisplay will NEVER be of any use for us! Correct me if I am wrong, please! :)

But I am sorry for putting you an that path in the first place and therefore "wasting" your time. That never was my intention.

On 26 July 2018 08:08:03 CEST, Quallenauge <notifications@github.com> wrote:
>1) Sure, there are issues. But to be honest, backlights in our devices
>have an operating range. Setting the brightness to 128 results in a
>different achieved effective brightness with your and my device. I
>think this can be tolerated, if the PWM sets the value to 128".1"/
>129".3333" or 127.9. The user won't notice that.
>
>3) You can do that of course. And this is an never-ending discussion
>right now. So I end here with this: In my opinion, the existing driver
>can be improved instead of trim down or re-implement.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lenovo-YTX703-Devel/android_kernel_lenovo_msm8976/pull/17#issuecomment-407987069
